### ISSUE TITLE
HashWithIndifferentAccess#select fails to return Enumerator

### ISSUE THREAD

**Issue by swifthand** [rails/rails#20095]
Commit 20c065594f1434c93e6fe64cad062a0df1f42a8e corrects the behavior of `HashWithIndifferentAccess#select` to be consistent with that of `Hash#select`, in that it returns the hash itself when unchanged. 

However, that fix introduced a different inconsistency with `Hash#select`, whereby HWIA's version will no longer return an `Enumerator` when no block is given. Thus `select` can no longer be added to with other Enumerator methods such as `with_index` and `with_object` and breaks expectations for anyone treating HWIA instances as full implementations of Enumerable.

A proposed solution follows after this demonstration of the issue.

**Example behavior for standard Hash**
```ruby
some_hash = { a: 1, b: 2, c: 3 }

# Hash#select behavior with block:
some_hash.select { |k, v| true }
# => {:a=>1, :b=>2, :c=>3}

# Hash#select behavior without block:
some_hash.select
#<Enumerator: {:a=>1, :b=>2, :c=>3}:select> 
some_hash.select.with_index
=> #<Enumerator: #<Enumerator: {:a=>1, :b=>2, :c=>3}:select>:with_index> 
```

**HWIA behavior before 20c065594f1434c93e6fe64cad062a0df1f42a8e**
```ruby
some_hwia = HashWithIndifferentAccess.new({ a: 1, b: 2, c: 3 })

# HWIA#select with block is inconsistent:
some_hwia.select { |k, v| true }
# => nil

# HWIA#select without block is consistent:
some_hwia.select
=> #<Enumerator: {:a=>1, :b=>2, :c=>3}:select>
some_hwia.select.with_index
=> #<Enumerator: #<Enumerator: {:a=>1, :b=>2, :c=>3}:select>:with_index>
```

**HWIA behavior after 20c065594f1434c93e6fe64cad062a0df1f42a8e**
```ruby
some_hwia = HashWithIndifferentAccess.new({ a: 1, b: 2, c: 3 })

# HWIA#select with block is now consistent:
some_hwia.select { |k, v| true }
# => {"a"=>1, "b"=>2, "c"=>3}

# HWIA#select without block is now inconsistent:
some_hwia.select
# => {"a"=>1, "b"=>2, "c"=>3}
some_hwia.select.with_index
# NoMethodError: undefined method `with_index' for {"a"=>1, "b"=>2, "c"=>3}:ActiveSupport::HashWithIndifferentAccess
```

The issue can be resolved, and both cases made consistent, if `HWIA#select` is implemented thusly:
```ruby
def select(*args, &block)
  copy = dup
  copy.select!(*args, &block) || copy
end
```

I do not immediately have a forked copy of Rails set up with which to provide tests and a pull request, but will endeavor to do so in the coming week. 

Please let me know if there is anything I can do to improve the description of this issue.

**Comment by swifthand** [comment]
**A note on use of truthiness:** In case there is concern about a possibility where calls to `Hash#select` should return false or nil (and thus making `|| copy` incorrect), I am not aware of any such behavior. A call to `Hash#select` that rejects every value will still return a new empty hash:
```ruby
{ a: 1, b: 2, c: 3 }.select { |k, v| false }
# => {}
HashWithIndifferentAccess.new({ a: 1, b: 2, c: 3 }).select { |k, v| false }
# => {}
```

**Comment by swifthand** [comment]
It looks like @imanel beat me to a pull request with #20125. 

Plus I think his use of `to_enum` is a better solution anyway!

**Comment by maclover7** [comment]
This should be closed since #20125 solves this issue, right?
